src/bin/nydusd/blob_cache.rs

```diff
@@ -277,28 +336,76 @@ impl BlobCacheMgr {
                 cache_compressed: false,
                 cache_config: entry.blob_config.cache_config.clone(),
                 cache_validate: false,
-                prefetch_config: match serde_json::from_value(entry.fs_prefetch.clone()) {
-                    Ok(fs_prefetch) => fs_prefetch,
-                    Err(_e) => Default::default(),
-                },
+                prefetch_config,
             },
         });
 
         Ok((path, factory_config))
     }
+
+    fn add_bootstrap_object(
+        &self,
+        domain_id: &str,
+        id: &str,
+        path: PathBuf,
+        factory_config: Arc<FactoryConfig>,
+    ) -> Result<()> {
+        let rs = RafsSuper::load_from_metadata(&path, RafsMode::Direct, true)?;
+        let bootstrap = BlobCacheObjectConfig::new_bootstrap_blob(
+            domain_id.to_string(),
+            id.to_string(),
+            path,
+            factory_config.clone(),
+        );
+
+        let mut state = self.get_state();
+        state.try_add(bootstrap.clone())?;
+        // Safe to unwrap() because it's a bootstrap.
+        let bs_obj = bootstrap.bootstrap_config().unwrap();
+
+        // Try to add the referenced data blob object if it doesn't exist yet.
+        for bi in rs.superblock.get_blob_infos() {
+            debug!(
+                "blob_cache: add data blob {} to domain {}",
+                &bi.blob_id(),
+                domain_id
+            );
+            let data_blob = BlobCacheObjectConfig::new_data_blob(
+                domain_id.to_string(),
+                bi,
+                factory_config.clone(),
+            );
+            let key = data_blob.get_key().to_string();
+
+            if let Err(e) = state.try_add(data_blob) {
+                // Rollback added bootstrap/data blobs.
+                let id = BlobCacheObjectId {
+                    domain_id: domain_id.to_string(),
+                    blob_id: id.to_string(),
+                };
+                let _ = state.remove(&id);
+                return Err(e);
+            }
+
+            // Associate the data blob with the bootstrap blob.
+            if let Some(BlobCacheObjectConfig::DataBlob(entry)) = state.id_to_config_map.get(&key) {
```

Comment:
We can do `data_blob.clone()` before this to simplify the code.

Reply:
Thanks, fixed.